src/coprocessor/dag/rpn_expr/types/expr_builder.rs

```diff
@@ -766,4 +805,126 @@ mod tests {
             );
         }
     }
+
+    #[test]
+    #[allow(clippy::iter_skip_next)]
+    fn test_expr_with_val() {
```

Comment:
Why not follow clippy?

BTW, the test looks strange to me. If we are just testing building an expression with `val`, one or two simple cases suffice. However, the biggest part of it is testing about **decode**.
If we really want to do so, I prefer move those parts to the datum mod. And that seems nothing to do with this PR, either. :)

Reply:
I just want to make sure we can get the corrected push down value. We will not use `datum::decode` to decode the pushed down datums in the future (TiDB will remove `Datum`).

Reply:
If TiDB will change the push-down format in the future, it seems we still won't benefit from hard-coded val as they also change...

Reply:
TiDB will not change the binary format, but the data will not be maintained as `Datum`. I mean we can change the `datum::decode` to `foo::decode` to make sure the result meets our expectations. BTW, I can remove it if you have strong opinions.

Reply:
I don't have strong opinions :)

Reply:
If the binary format isn't expected to change, it looks fine to me.